### PR TITLE
[CFF2] Support ‘maxstack’

### PR DESCRIPTION
**Pull request by miguelsousa**
A late patch from @readroberts.

### CODE REVIEW

File: Lib/fontTools/cffLib.py [hunk 1]
```diff
@@ -414,6 +417,7 @@ def getItems(self, items, strings):
 
 
 class TopDictIndexCompiler(IndexCompiler):
+	maxOpStack = 193
```
Comment (anthrotype) [inline Lib/fontTools/cffLib.py]: @miguelsousa where is this attribute used? why is it not not simply `maxOpStack = defaultMaxStack`?

File: Lib/fontTools/cffLib.py [hunk 2]
```diff
@@ -1731,17 +1743,16 @@ def xmlRead(self, name, attrs, content, parent):
 	((12, 36),	'FDArray',		'number',	None,	FDArrayConverter()),
 	(17,		'CharStrings',		'number',	None,	CharStringsConverter()),
 	(24,		'VarStore',		'number',	None,	VarStoreConverter()),
-	(25,		'maxstack',		'number',	None,	None),
```
Comment (anthrotype) [inline Lib/fontTools/cffLib.py]: any specific reason why you moved this `maxstack` operator at the top of this list? just curious

Comment (readroberts) [inline Lib/fontTools/cffLib.py]: Yes.  maxstack needs to be set before any PrivateDict is processed, in order to be able to parse the blend operators correctly. The simplest way to do this is to put maxstack first in the order, although it only needs to be before the FDArray.

File: Lib/fontTools/cffLib.py (hunk 1 above)
Comment (readroberts) [inline Lib/fontTools/cffLib.py]: It should indeed be maxOpStack = defaultMaxStack.

File: Lib/fontTools/cffLib.py (hunk 2 above)
Comment (anthrotype) [inline Lib/fontTools/cffLib.py]: I wonder, if this `maxstack` is a CFF2-only operator why is it also listed in `topDictOperators` and not just in `topDictOperatos2`?
I see it's mentioned in CFF2 spec https://www.microsoft.com/typography/otspec/cff2charstr.htm, but don't see any reference to it in the CFF 1.0 spec (Tech Note #5176).

Comment (readroberts) [inline Lib/fontTools/cffLib.py]: This a work-around for a current limitation on my scripts for building OpenType/CFF2 fonts. It is currently convenient to able to convert the default master CFF object to a CFF2 object by doing a bit of field editing and then compiling it. That requires the CFF compiler methods to be able to handle the CFF2 fields. This is messy, and I plan to fix this in the longer run - it is not particularly difficult to do so.

Comment (readroberts) [inline Lib/fontTools/cffLib.py]: More specifically, I plan to work in this in about two weeks, when I will also remove the global data references.

**Comment by behdad** [comment]
> I see it's mentioned in CFF2 spec https://www.microsoft.com/typography/otspec/cff2charstr.htm, but don't see any reference to it in the CFF 1.0 spec (Tech Note #5176).

I should have raised this way before, but I truly don't see why we need a maxstack operator.  I'm sure all implementations could have just used the max maxstack and be fine. Bummer

File: Lib/fontTools/cffLib.py (hunk 1 above)
Comment (miguelsousa) [inline Lib/fontTools/cffLib.py]: Fixed

**Comment by anthrotype** [comment]
> I should have raised this way before, but I truly don't see why we need a maxstack operator. I'm sure all implementations could have just used the max maxstack and be fine.

Yes, but it's in the official spec by now so I believe we should support it until/unless is officially deprecated.

> I plan to work in this in about two weeks, when I will also remove the global data references.

Thanks @readroberts.

Read, Miguel, you must have been using this updated cffLib for a while internally; do you believe this is stable enough to cut a new fonttools release?

I'd like to push out some varLib bugfixes today so that fontmake use of them. However, I fear that after merging `cff2` branch we may have introduced other regressions in the CFF 1.0 part of the code which we might not be aware of.

I guess it's a bit too late to revert the ``cff2`` merge; so I'm thinking of tagging 3.9.0 and wish that users will spot and report back any issues they may find.

**Comment by miguelsousa** [comment]
@anthrotype it's as stable as it can be for the time being, until @readroberts has time to clean up the code. So I'd say go ahead with a new release. Would be great if the #881 changes were included as well, but it sounds like they will have to wait until they're validated.

**Pull request merged**